### ISSUE TITLE
`Queue::fake()` does not fully mimic a real job `dispatch()`

### ISSUE THREAD

**Issue by bilogic** [laravel/framework#37509]
- Laravel Framework 8.44.0
- PHP 7.4.13 (cli) (built: Nov 24 2020 10:03:34) ( NTS )
- mysql  Ver 8.0.21 for Linux on x86_64 (Source distribution)

### Description:

`Queue::fake()` does not fully mimic a real job `dispatch()`.
When `$head` is `protected`, the unserialized class' `$head` is empty, even though `$head` has some value in the `jobs` table entry.

### Steps To Reproduce:

```PHP
class myjob implements ShouldQueue
{
    use Queueable, SerializesModels, InteractsWithQueue, Dispatchable;

    protected $head = [];

    public function getHead()
    {
        return $this->head;
    }

    public function setHead($key, $value)
    {
        $this->head[$key] = $value;

        return $this;
    }
}

public function test_Can_set_head()
{
    Queue::fake();

    $mailer = app(myjob::class)
        ->setHead('key', 'value')
        ->dispatch();

    Queue::assertPushed(myjob::class, function ($job) {
        return $job->getHead()['key'] == 'value';
    });
}
```

1. The test above will pass.
2. However, in a real database job `dispatch()`
    - `$head['key']` will not be set
    - It is only when `$head` is changed to `public`, `key` gets set with `value`
3. Regardless of `public`/`protected`, I could see `key = value` in the `jobs` table, so perhaps it could be that `dispatch()` does not behave the same as `Queue::fake()`

**Comment by driesvints** [comment]
Please post the full code to reproduce this as `setHead` doesn't exists in Laravel.

**Comment by bilogic** [comment]
Sure, I have updated the OP

**Comment by driesvints** [comment]
So I think you misunderstand a bit how `->dispatch()` works. That's actually not a proper usage of the method. It's meant as a static method call on an uninitialised job class. It accepts the arguments that are passed to the constructor of the job class. Behind the hood it'll initialise a new job instance and thus it'll loose all state it had (from the `setHead` call).

Try:

```php
<?php

namespace App\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;

class FooJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $head = [];

    public function __construct(array $head = [])
    {
        $this->head = $head;
    }

    public function getHead()
    {
        return $this->head;
    }
}
```

```php
<?php

namespace Tests\Feature;

use App\Jobs\FooJob;
use Illuminate\Support\Facades\Queue;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    public function testBasicTest()
    {
        Queue::fake();

        FooJob::dispatch(['key' => 'value']);

        Queue::assertPushed(FooJob::class, function ($job) {
            return $job->getHead()['key'] == 'value';
        });
    }
}
```

And you'll see your tests pass. In general I'd like to say that it's a best practice to never use setters on a job and always pass all state needed to handle the job through its constructor.

**Comment by bilogic** [comment]
@driesvints 

What you shared is new to me, I will give it a test shortly, however, I wanted to make sure my points are clear:

1. The **test in my OP is passing**, not failing
2. Thus, I took the test code and put it into production
3. However, it does not work as expected in production, the job gets dispatched etc, but some of the values are wrong
4. What I did to make it work in production was to change `protected $head` to `public $head`
5. So perhaps the question here is, can `$head` be `protected` or must it be `public`?
6. If it must be `public`, why does the OP test pass when it is `protected`?

Thank you.
